### PR TITLE
GH-34157: [C++] Configure bundled AWS SDK to use aws-lc instead of OpenSSL

### LINKED ISSUE #34157: [C++] Configure bundled AWS SDK to use aws-lc instead of OpenSSL
### Describe the enhancement requested

Using OpenSSL causes various issues like https://github.com/apache/arrow/pull/33808#issuecomment-1408247269 and https://github.com/apache/arrow/issues/34111. We should try to use aws-lc for libcrypto and libssl. We need to hide them inside s2n-tls to avoid name conflicts with OpenSSL used by other libraries.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by js8544**
Using OpenSSL causes various issues like https://github.com/apache/arrow/pull/33808#issuecomment-1408247269 and https://github.com/apache/arrow/issues/34111. We should try to use aws-lc for libcrypto and libssl. We need to hide them inside s2n-tls to avoid name conflicts with OpenSSL used by other libraries.

Some notes:
1. Only linux needs s2n-tls and aws-lc.
2. Because aws-c-http requires curl which requires libcrypto from OpenSSL, we need to hide Aws::libcrypto by S2N_INTERN_LIBCRYPTO=ON. 
3. aws-c-cal also uses libcrypto, but it can't hide it. So we need to use OpenSSL instead of aws-lc for it.
4. Mac and windows don't need s2n-tls, but do need some security libs from system, like `-framework Security` and `ncrypt.lib`. 
5. arrow-s3fs-test is re-enabled.
6. Force Java Gandiva to use static protobuf lib.

* Closes: #34157

### CODE REVIEW

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -4727,20 +4735,38 @@ macro(build_awssdk)
...
+      else()
+        execute_process(COMMAND ${BREW} --prefix "openssl"
+                        OUTPUT_VARIABLE OPENSSL_BREW_PREFIX
+                        OUTPUT_STRIP_TRAILING_WHITESPACE)
+        if(OPENSSL_BREW_PREFIX)
+          set(OPENSSL_ROOT_HINT ${OPENSSL_BREW_PREFIX})
+        endif()
+      endif()
+    endif()
+  endif()
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4757]: Do we need this?
`OPENSSL_INCLUDE_DIR` must be set by `find_package(OpenSSL)`.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -4827,37 +4881,41 @@ macro(build_awssdk)
...
-        if(OPENSSL_BREW_PREFIX)
-          set(OPENSSL_ROOT_DIR ${OPENSSL_BREW_PREFIX})
-        endif()
-      endif()
-    endif()
-  endif()
-  if(OPENSSL_ROOT_DIR)
-    # For Findcrypto.cmake in s2n-tls.
-    list(APPEND S2N_TLS_CMAKE_ARGS -DCMAKE_PREFIX_PATH=${OPENSSL_ROOT_DIR})
+  if(UNIX AND NOT APPLE) # aws-lc and s2n-tls only needed on linux
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4884]:
```suggestion
  if("s2n-tls" IN_LIST _AWSSDK_LIBS)
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -4827,37 +4881,41 @@ macro(build_awssdk)
...
-                      ${EP_COMMON_OPTIONS}
-                      URL ${S2N_TLS_SOURCE_URL}
-                      URL_HASH "SHA256=${ARROW_S2N_TLS_BUILD_SHA256_CHECKSUM}"
-                      CMAKE_ARGS ${S2N_TLS_CMAKE_ARGS}
-                      BUILD_BYPRODUCTS ${S2N_TLS_STATIC_LIBRARY})
-  add_dependencies(AWS::s2n-tls s2n_tls_ep)
 
+  set(AWS_C_CAL_CMAKE_ARGS ${AWSSDK_COMMON_CMAKE_ARGS})
+  # Because aws-c-cal doesn't provide ability to hide libcrypto, we have to always use openssl
+  list(APPEND AWS_C_CAL_CMAKE_ARGS -DUSE_OPENSSL=ON)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4918]: Could you use this in `externalproject_add(aws_c_cal_ep)`?
Or is it needless?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -4867,13 +4925,17 @@ macro(build_awssdk)
                       DEPENDS aws_c_common_ep)
   add_dependencies(AWS::aws-c-cal aws_c_cal_ep)
 
+  set(AWS_C_IO_DEPENDS aws_c_common_ep aws_c_cal_ep)
+  if(UNIX AND NOT APPLE)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:4929]:
```suggestion
  if(TARGET s2n_tls_ep)
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 5]
```diff
@@ -4985,6 +5047,14 @@ macro(build_awssdk)
     set_property(TARGET CURL::libcurl
                  APPEND
                  PROPERTY INTERFACE_LINK_LIBRARIES OpenSSL::SSL)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:5049]: This is not related to this pull rquest but we should do this in `find_curl()`.

**Review by kou — COMMENTED** [review]
Thanks!

File: ci/scripts/java_jni_macos_build.sh [hunk 6]
```diff
@@ -135,6 +136,7 @@ archery linking check-dependencies \
   --allow libncurses \
   --allow libobjc \
   --allow libplasma_java \
+  --allow libprotobuf \
```
[the remark this answers is not in the document]
Comment (js8544) [reply ci/scripts/java_jni_macos_build.sh:139]: https://github.com/ursacomputing/crossbow/actions/runs/4165692619/jobs/7209047999 From the error log: 

`2023-02-13T17:54:24.6436120Z Error: Unexpected shared dependency found in libgandiva_jni.dylib: libprotobuf`

(You need to view raw log to see this. The log size is too large.)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4757]: I copied it from there: https://github.com/apache/arrow/pull/34159/files#diff-5cdc95f4e1b618f2f3ef10d370ce05a1ac05d9d401aecff3ccbb3d76bd366b6aL4831. Was it needed in the previous PR?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 3 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4918]: Seems like it's not needed.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 7]
```diff
@@ -4796,6 +4835,11 @@ macro(build_awssdk)
           "${AWSSDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}s2n${CMAKE_STATIC_LIBRARY_SUFFIX}"
       )
     endif()
+    if(${_AWSSDK_LIB} STREQUAL "crypto" OR ${_AWSSDK_LIB} STREQUAL "ssl")
```
[the remark this answers is not in the document]
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4838]: changed to aws-lc

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 8]
```diff
@@ -4779,9 +4805,22 @@ macro(build_awssdk)
...
+  set(AWS_LC_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/aws_lc_ep-install")
+  set(AWS_LC_INCLUDE_DIR "${AWS_LC_PREFIX}/include")
+
+  if(UNIX AND NOT APPLE) # aws-lc and s2n-tls only needed on linux
+    file(MAKE_DIRECTORY ${AWS_LC_INCLUDE_DIR})
+    list(APPEND
+         _AWSSDK_LIBS
+         s2n-tls
+         crypto
+         ssl)
```
[the remark this answers is not in the document]
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4821]: done

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4757]: aws-c-cal will do a find_package(OpenSSL), but i'm not sure if it needs this though

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 5 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Moved to find_curl

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 3 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Removed.

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

File: ci/scripts/java_jni_macos_build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (js8544) [reply ci/scripts/java_jni_macos_build.sh:139]: I'm confused...... IIRC Gandiva doesn't need protobuf. https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L367 This should be removed.

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

File: ci/scripts/java_jni_macos_build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (js8544) [reply ci/scripts/java_jni_macos_build.sh]: From what I can tell, [This line](https://github.com/apache/arrow/blob/master/java/gandiva/CMakeLists.txt#L41) gives this: 
```
2023-02-14T08:45:58.6020450Z -- Found Protobuf: /usr/local/lib/libprotobuf.dylib (found version "3.21.12") 
```
in [this log](https://pipelines.actions.githubusercontent.com/serviceHosts/36b0b1a0-9453-4f79-b171-2b3905790445/_apis/pipelines/1/runs/134770/signedlogcontent/7?urlExpires=2023-02-14T08%3A59%3A03.1286297Z&urlSigningMethod=HMACV1&urlSignature=LtsaqA2TGb81lINJYHSn9P7YcBMFO89zYknLi%2Bik4LY%3D). Is it expected?

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

File: ci/scripts/java_jni_macos_build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (js8544) [reply ci/scripts/java_jni_macos_build.sh]: I've changed Java Gandiva to use static protobuf lib. Let's see if it solves the problem.

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
CMake somehow cannot find brew installed libprotobuf even if I supply Protobuf_ROOT. [The error log](https://pipelines.actions.githubusercontent.com/serviceHosts/36b0b1a0-9453-4f79-b171-2b3905790445/_apis/pipelines/1/runs/134813/signedlogcontent/7?urlExpires=2023-02-14T12%3A24%3A29.0803195Z&urlSigningMethod=HMACV1&urlSignature=Vq8uKR48GIOK0zVJXSsnz27XHch8h5UveMA3dqRsXds%3D) gives:
```
2023-02-14T11:31:53.4607330Z -- Could NOT find Protobuf (missing: Protobuf_LIBRARIES) (found suitable version "3.21.12", minimum required is "2.6.1")
2023-02-14T11:31:53.4610530Z -- Building Protocol Buffers from source
```

I checked the [last successful java-jars build](https://github.com/ursacomputing/crossbow/actions/runs/4122242694/jobs/7118788009
), it does find brew's protobuf correctly:
```
Found Protobuf: /usr/local/lib/libprotobuf.a (found suitable version "3.21.12", minimum required is "2.6.1") 
```

Everything else is working except for this protobuf issue which is actually unrelated to this pr. cc @kou Could you have a look?

**Comment by kou** [comment]
Ah, this is caused by https://github.com/Homebrew/homebrew-core/pull/122277 .
Homebrew stopped providing `libprotobuf.a`.

We need to prepare `libprotobuf.a` for us. But it is not related to this pull request. Could you open a new issue for it?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:4756]: It was needed but I should have reused `OPENSSL_ROOT_HINT`. I didn't notice it. Sorry.
Could you try removing this?

File: java/gandiva/CMakeLists.txt [hunk 9]
```diff
@@ -38,6 +38,7 @@ set(GANDIVA_PROTO_DIR ${CMAKE_CURRENT_SOURCE_DIR}/proto)
 get_filename_component(GANDIVA_PROTO_FILE_ABSOLUTE ${GANDIVA_PROTO_DIR}/Types.proto
                        ABSOLUTE)
 
+set(Protobuf_USE_STATIC_LIBS ON)
```
Comment (kou) [inline java/gandiva/CMakeLists.txt:41]: Could you specify this in build script (`ci/scripts/java_jni_build.sh`)?
Some users may want to use `libprotobuf.so`.

File: java/gandiva/CMakeLists.txt [hunk 10]
```diff
@@ -38,6 +38,7 @@ set(GANDIVA_PROTO_DIR ${CMAKE_CURRENT_SOURCE_DIR}/proto)
...
 find_package(Protobuf REQUIRED)
 if(MSVC)
```
Comment (kou) [inline java/gandiva/CMakeLists.txt:43]: It's not related to this pull request but could you remove this `if(MSVC)`? It's not needed with recent CMake.

File: java/gandiva/CMakeLists.txt (hunk 9 above)
Comment (js8544) [reply java/gandiva/CMakeLists.txt]: Done.

File: java/gandiva/CMakeLists.txt (hunk 10 above)
Comment (js8544) [reply java/gandiva/CMakeLists.txt]: Done.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (js8544) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Removed.

**Comment by js8544** [comment]
> Ah, this is caused by [Homebrew/homebrew-core#122277](https://github.com/Homebrew/homebrew-core/pull/122277) . Homebrew stopped providing `libprotobuf.a`.
> 
> We need to prepare `libprotobuf.a` for us. But it is not related to this pull request. Could you open a new issue for it?

Nice find! I thought of this reason but didn't find this change. I created a new issue https://github.com/apache/arrow/issues/34189 for this.

**Comment by js8544** [comment]
CI failures are unrelated. I think this is ready to to merged. cc @kou

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g r java-jars

**Comment by js8544** [comment]
@github-actions crossbow submit -g cpp -g r java-jars

**Comment by kou** [comment]
Sorry. Can we try this?

```diff
diff --git a/ci/scripts/java_jni_macos_build.sh b/ci/scripts/java_jni_macos_build.sh
index 391d05642..37bbe1283 100755
--- a/ci/scripts/java_jni_macos_build.sh
+++ b/ci/scripts/java_jni_macos_build.sh
@@ -109,6 +109,7 @@ fi
 popd
 
 
+export JAVA_JNI_CMAKE_ARGS="-DProtobuf_ROOT=${build_dir}/cpp/protobuf_ep-install"
 ${arrow_dir}/ci/scripts/java_jni_build.sh \
   ${arrow_dir} \
   ${install_dir} \
```

**Comment by js8544** [comment]
@github-actions crossbow submit java-jars

**Comment by js8544** [comment]
> Sorry. Can we try this?
> 
> ```diff
> diff --git a/ci/scripts/java_jni_macos_build.sh b/ci/scripts/java_jni_macos_build.sh
> index 391d05642..37bbe1283 100755
> --- a/ci/scripts/java_jni_macos_build.sh
> +++ b/ci/scripts/java_jni_macos_build.sh
> @@ -109,6 +109,7 @@ fi
>  popd
>  
>  
> +export JAVA_JNI_CMAKE_ARGS="-DProtobuf_ROOT=${build_dir}/cpp/protobuf_ep-install"
>  ${arrow_dir}/ci/scripts/java_jni_build.sh \
>    ${arrow_dir} \
>    ${install_dir} \
> ```

It works. Thanks for suggesting this!

**Review by kou — APPROVED** [review]
+1

Thanks!

**Comment by kou** [comment]
@github-actions crossbow submit -g wheel

**Comment by kou** [comment]
@js8544 It seems that we need more work for manylinux wheels. Could you open a new issue for it?

**Pull request merged**